File: src/com/gmail/marcosav2010/tasker/Tasker.java
package com.gmail.marcosav2010.tasker;

import java.util.Collections;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import com.gmail.marcosav2010.common.Utils;

public class Tasker {

	private final Object lock = new Object();
	private final AtomicInteger taskCounter = new AtomicInteger();
	private final Map<Integer, Task> tasks = new ConcurrentHashMap<>();
	private final Map<TaskOwner, Set<Task>> tasksByOwner = new ConcurrentHashMap<>();

	public void cancel(int id) {
		Task task = tasks.get(id);
		if (task == null)
			throw new IllegalArgumentException("No task with id " + id);

		task.cancel();
	}

	void cancel0(Task task) {
		synchronized (lock) {
			tasks.remove(task.getId());
			Utils.remove(tasksByOwner, task);
		}
	}

	public void cancel(Task task) {
		task.cancel();
	}

	public int cancel(TaskOwner peer) {
		Set<Task> toRemove = Collections.synchronizedSet(new HashSet<>());
		
		tasksByOwner.get(peer).forEach(toRemove::add);

		toRemove.forEach(this::cancel);
		return toRemove.size();
	}

	public Task run(TaskOwner owner, Runnable task) {
		return schedule(owner, task, 0, TimeUnit.MILLISECONDS);
	}

	public Task schedule(TaskOwner owner, Runnable task, long delay, TimeUnit unit) {
		return schedule(owner, task, delay, 0, unit);
	}

	public Task schedule(TaskOwner owner, Runnable task, long delay, long period, TimeUnit unit) {
		if (owner == null)
			throw new IllegalArgumentException("Owner cannot be null ");
		if (task == null)
			throw new IllegalArgumentException("Task runnable cannot be null ");
		
		Task prepared = new Task(this, owner, taskCounter.getAndIncrement(), task, delay, period, unit);

		synchronized (lock) {
			tasks.put(prepared.getId(), prepared);
			Utils.put(tasksByOwner, owner, prepared);
		}

		owner.getExecutorService().execute(prepared);
		return prepared;
	}
}

File: core/src/main/java/com/gmail/marcosav2010/tasker/Tasker.kt
package com.gmail.marcosav2010.tasker

import com.gmail.marcosav2010.common.Utils
import java.util.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger
import java.util.function.Consumer

object Tasker {

    private val lock = Any()
    private val taskCounter = AtomicInteger()
    private val tasks: MutableMap<Int, Task> = ConcurrentHashMap()
    private val tasksByOwner: MutableMap<TaskOwner, MutableSet<Task>> = ConcurrentHashMap()

    fun cancel(id: Int) = (tasks[id] ?: throw IllegalArgumentException("No task with id $id")).cancel()

    internal fun cancel0(task: Task) {
        synchronized(lock) {
            tasks.remove(task.id)
            Utils.remove(tasksByOwner, task)
        }
    }

    private fun cancel(task: Task) = task.cancel()

    fun cancel(peer: TaskOwner): Int {
        val toRemove = Collections.synchronizedSet(HashSet<Task>())
        toRemove.addAll(tasksByOwner[peer] ?: error(""))
        toRemove.forEach(Consumer { task -> cancel(task) })
        return toRemove.size
    }

    fun run(owner: TaskOwner, task: () -> Unit) = schedule(owner, task, 0)

    fun schedule(owner: TaskOwner, task: () -> Unit, delay: Long, unit: TimeUnit) = schedule(owner, task, delay, 0, unit)

    fun schedule(owner: TaskOwner,
                 task: () -> Unit,
                 delay: Long,
                 period: Long = 0,
                 unit: TimeUnit = TimeUnit.MILLISECONDS): Task {

        val prepared = Task(this, owner, taskCounter.getAndIncrement(), task, delay, period, unit)

        synchronized(lock) {
            tasks[prepared.id] = prepared
            Utils.put(tasksByOwner, owner, prepared)
        }

        owner.executorService.execute(prepared)
        return prepared
    }
}